tests/jsrun.py

```diff
@@ -88,7 +89,7 @@ def require_engine(engine):
     sys.exit(1)
 
 
-def run_js(filename, engine=None, args=[],
+def run_js(filename, engine=None, args=[], command_flags=None,
```

Comment:
are the changes to this file needed? I'm not sure where they are used in the diff.

Reply:
I used it as part of development, debugging the `test_assertions_on_ready_promise`, by being able to pass `--inspect-brk` to the `run_js` call [here](https://github.com/emscripten-core/emscripten/pull/11615/files#diff-9ee3e07b8e34e901a62806f99766f262R9986). I thought it was useful enough to keep in. Let me know if you'd like to see that tip included in development docs somewhere.

Reply:
I'd rather you revert this part, at least until we consider it a little more.

I'm doing a bunch of refactoring around run_js at the moment and trying to reduce the complexity and number of args it takes.

Reply:
Sure, reverted. Please make sure the generated JS for tests are easily debuggable, workflow wise, in the improved version you're working on. It's been pretty difficult having to learn how the entire test runner is put together when I just want to fix one or two failing tests.

Reply:
It sounds like you talking about added extra nodejs flags?

One method that myself and other employ heavily during debugging is `EMTEST_SAVE_DIR=1` this will allow you to go in an see the test data, and re-run the JS code outside of the test runner.   If you combine with EMCC_DEBUG=1 you can see all the commands that you might want to re-run.  

Its very useful for iterating on a given test quicky.

If you have specific examples of ways that you would prefer things to work feel free to propose them.

Reply:
Yup, extra nodejs flags. I've been using those flags as well, but it was nicer to just add a parameter of `command_flags=['--inspect-brk']` and being able to debug the running test rather than having to copy and paste commands and execute them myself in whatever directory `EMTEST_SAVE_DIR` went to (which on macOS is a random temp directory).

Reply:
Yeah I think moving the test temp directories inside the emscripten directory itself is one thing that might be better here.

On linux its always `/tmp/emscripten_test`..   I did notice it was different on macOS .. I guess they do TMPDIR differently there?

Reply:
In fact was thinking of not treating test output as temporary at all and having a permanent output directory for each test.    e.g. `emscripten/tests/out/test_doo/...`     I guess it would end up being very big after full test run.. but would certainly aid debugging. 

This is how the llvm testsuite runs I beleive.

Reply:
Yeah, it's different on macOS. Here's an [article about it](https://osxdaily.com/2018/08/17/where-temp-folder-mac-access/).

Having the output be clearly organized and easily accessible is nice, but it still leaves the obstacle of knowing what command to execute on the output when debugging. Attaching the debugger to a running test has been the best tool for me to figure out what's happening.

Relatedly, it would also be nice to be able to easily specify something to automatically open the browser's devtools when doing browser tests, i.e. Chrome's `--auto-open-devtools-for-tabs` flag and Firefox's `--jsconsole` flag. As it stands now, you have to quickly open devtools yourself when the browser opens; I often lose that race, the test finishes, and I've gotta re-run it. I know it's possible to do so through `EMTEST_BROWSER`, it's just a pain and kind of difficult to figure out.

All of this could just be a matter of better development documentation.

Reply:
When you say "Attaching the debugger " do you mean some kind of JS debugger?  Or gdb?

Sounds useful either way.

I think the best way to modify node flags in this way would be via the config file, or perhaps via the environment?  `EMTEST_JS_ARGS=XXX` maybe?  

For the config file option, you probably don't want to set NODE_JS itself since that is also used by emcc, not just by the test suite.  However you cat set `JS_ENGINES=[...]` in the `.emscripten` and this will only effect the test suite.

Reply:
Right, I'm referring to the JS debugger, not gdb or pdb.

Those might work if I want to debug the whole suite, but if there's a single test I want to debug I find the best balance between convenience and straightforwardness to be able to easily do so from the test's source code. Using environment flags or configuration files requires knowledge of their existence and effect, whereas the source code is WYSIWYG.